### PR TITLE
[PUBDEV-4373] JMH benchmarks for PCA

### PR DESCRIPTION
**Pull request by mathemage**
For [the assignment to improve H2O PCA](https://gist.github.com/mmalohlava/717ad7b7441a6ff91b5f0a907482bd5d), JMH benchmarks for PCA need to be written.

Do not merge! Still **work in progress**!!

### CODE REVIEW

**Comment by mathemage** [comment]
@mmalohlava @arnocandel @michalkurka @jakubhava I couldn't figure out how to add labels to PRs without write access. Can you label this PR with **WIP** and **please review** labels, please? And also add yourselves as reviewers, if you wish?

**Comment by mmalohlava** [comment]
@h2o-ops please test!

**Comment by mathemage** [comment]
@mmalohlava @michalkurka 
How do you run JMH benchmarks in H2O? I usually do simply 
```
./gradlew clean
./gradlew :h2o-algos:jmh
```
but I just found a file named `h2o-3/h2o-core/src/jmh/java/water/jmh/H2oJmhRunner.java`!

**Comment by mmalohlava** [comment]
Hi Karel,

running `:h2o-core:jmh` is fine, there is also task `:h2o-core:ubench` which is using `H2OJmhRunner` but the main reason for runner is to extract info about run and save it into csv file.

**Comment by mathemage** [comment]
@mmalohlava Thanks for explanation, this really simplifies CSV extraction for me! :happy:

**Comment by mmalohlava** [comment]
Hi Karel,

do you have any preliminary results comparing JAMA and JNI?

Two additional notes:
  - for final PR, you will need to remove datasets
  - if JMH fails for you, a simple measurement with `System.currentTimeMillis` could be enough

**Comment by mathemage** [comment]
Hi Michal,

I am still working on my assignment. I have implemented JNI (MTJ) SVD in this [new feature branch](https://github.com/h2oai/h2o-3/compare/master...mathemage:mathemage-pca-using-netlib-java-svd). The new version passes the PCA JUnit tests, so it should be fine. 

I haven't measured them yet, though, since I've been struggling with making JMH to work. There are some obstacles with threads, cluster, nullPointerExceptions and what-not-runtime-problems. You know, my first time with JMH :smile: I'll poke around with JMH for a while and see if I can push it through. Otherwise I'll try your `System.currentTimeMillis` idea soon!

As for removing datasets, do you suggest any neat solution? E.g. sync data files with Gradle somehow?

Thanks for all your hints!

**Comment by mathemage** [comment]
@mmalohlava @michalkurka `./gradlew :h2o-algos:jmh` gives me error:
```
Jar check fails; my hash=[-45, 36, -10, 118, 6, 119, -55, -65, -14, 97, 91, -107, 64, -45, -71, 86]
Jar check fails; received hash=[124, -53, 37, 73, 2, 126, -24, -110, 79, -7, -52, 115, 104, -35, -118, 81]
04-04 00:03:33.988 10.76.97.14:54323     31964  FJ-126-7  FATAL: Attempting to join /10.76.97.14:54321 with an H2O version mismatch (md5 differs).  (Is H2O already running?)  Exiting.
2017-04-04 00:03:34.001:INFO:oejsh.ContextHandler:stopped o.e.j.s.ServletContextHandler{/,null}
```
I tried to Google, nothing so far. Any ideas how to fix this? Perhaps killing h2o cluster between iterations?

**Comment by mathemage** [comment]
@mmalohlava BTW to answer your question on preliminary numbers between JAMA and JNI, the JUnit tests seemed run for JNI (9 s - 18 s) than for Jama (11 s - 20 s). But that was just manual guess and unit tests are not really precise. That's why I want to verify with JMH.

**Comment by mathemage** [comment]
@mmalohlava 
> for final PR, you will need to remove datasets

How about creating a symlink from `resources/smalldata` via a Gradle task? Or how is it done with resources for other JMH benchmarks in H2O-3?

**Comment by mathemage** [comment]
> do you have any preliminary results comparing JAMA and JNI?

Some preliminary test results for new JNI implementation:
```
Benchmark                                           (dataSetCase)  Mode  Cnt   Score   Error  Units
PCABenchSuite.measureImputeMissingTraining              1  avgt   10   5.516 ± 3.000  ms/op
PCABenchSuite.measureWideDataSetsBenchScoringCase       1  avgt   10   0.312 ± 0.085  ms/op
PCABenchSuite.measureWideDataSetsBenchTrainingCase      1  avgt   10  17.882 ± 5.988  ms/op
```

**Comment by mmalohlava** [comment]
@mathemage cool! do you have comparison with JAMA-based implementation?

**Comment by mathemage** [comment]
@mmalohlava nope, I need to make JMH work properly, it still crashes. Then I finalize this PR and add the benchmarks as a new feature and then I run JMH on both implementations. Sorry for taking so long, my laptop RAM size is failing me in JMH runs :sob:

**Comment by mathemage** [comment]
Hi everyone!

I am getting this exception during JMH benchmarks:
```
java.lang.RuntimeException: Rollups not possible, because Vec was deleted: $04ff0e000000ffffffff$nfs:/../smalldata/pca_test/decathlon.csv
	at water.fvec.RollupStats.get(RollupStats.java:321)
	at water.fvec.RollupStats.get(RollupStats.java:348)
	at water.fvec.Vec.rollupStats(Vec.java:816)
	at water.fvec.Vec.mins(Vec.java:739)
	at water.fvec.Vec.min(Vec.java:736)
	at water.fvec.Vec.isConst(Vec.java:748)
	at hex.ModelBuilder$3.filter(ModelBuilder.java:699)
	at hex.ModelBuilder$FilterCols.doIt(ModelBuilder.java:1133)
	at hex.ModelBuilder.ignoreBadColumns(ModelBuilder.java:700)
	at hex.ModelBuilder.init(ModelBuilder.java:862)
	at hex.pca.PCA.init(PCA.java:94)
	at hex.pca.PCA$PCADriver.computeImpl(PCA.java:239)
	at hex.ModelBuilder$Driver.compute2(ModelBuilder.java:170)
	at water.H2O$H2OCountedCompleter.compute(H2O.java:1202)
	at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)
	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)
	at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:974)
	at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1477)
	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)
```

Do you have any ideas where the problem is?

**Comment by mmalohlava** [comment]
@mathemage You need to decrease parallelization factor for your benchmark - my theory is that JMH will run benchmark in parallel and some of underlying data structures will get inconsistent.

**Comment by mmalohlava** [comment]
From JMH

```
@Threads	Number of threads to use for the test. The default is Runtime.getRuntime().availableProcessors().
```

**Comment by mathemage** [comment]
@mmalohlava 
Thanks for the tip on threads, it helped! I can run benchmarks on the Jama-based version now:
```
Benchmark                                                   Mode  Cnt  Score   Error  Units
PCAImputeMissingBenchTraining.measureImputeMissingTraining  avgt   10  2.291 ± 1.837  ms/op
PCAImputeMissingScoringBench.measureImputeMissingScoring    avgt   10  0.311 ± 0.287  ms/op
```

**Comment by mathemage** [comment]
I am trying to use the same JMH benchmarks on the [JNI-based implementation of PCA](https://github.com/mathemage/h2o-3/tree/mathemage-pca-using-netlib-java-svd), however, I am running into this [issue](http://stackoverflow.com/questions/32361545/gradle-trying-to-unzip-a-pom-typed-dependency). I'll fiddle with Maven dependencies for a while to see how to fix it...

**Comment by mathemage** [comment]
> I am trying to use the same JMH benchmarks on the JNI-based implementation of PCA, however, I am running into this issue. I'll fiddle with Maven dependencies for a while to see how to fix it...

:ok: , I resolved the issue above :+1: and ran first benchmarks of both PCA versions:

### Jama
```
Benchmark                                                   Mode  Cnt  Score   Error  Units
PCAImputeMissingBenchTraining.measureImputeMissingTraining  avgt   10  2.312 ± 0.911  ms/op
PCAImputeMissingScoringBench.measureImputeMissingScoring    avgt   10  0.436 ± 0.536  ms/op
```

### MTJ (netlib-java)
```
Benchmark                                                   Mode  Cnt  Score   Error  Units
PCAImputeMissingBenchTraining.measureImputeMissingTraining  avgt   10  2.247 ± 0.997  ms/op
PCAImputeMissingScoringBench.measureImputeMissingScoring    avgt   10  0.624 ± 0.846  ms/op
```

### System info
```bash
$ inxi 
CPU~Dual core Intel Core i5-4300M (-HT-MCP-) speed/max~2485/3300 MHz Kernel~4.4.0-72-generic x86_64 Up~3:37 Mem~2390.6/3831.1MB HDD~500.1GB(50.1% used) Procs~247 Client~Shell inxi~2.2.35  

$ inxi -S
System:    Host: ThinkPad-T440p Kernel: 4.4.0-72-generic x86_64 (64 bit) Desktop: Cinnamon 2.8.6
           Distro: Ubuntu 16.04 xenial
```

### Conclusion
Not very flattering for MTJ, it's not really a drastic improvent. Maybe it's because of the simplicity of the benchmarks, so I'll create more benchmarks based on the JUnit tests. I'll let you know then.

**Comment by mathemage** [comment]
@mmalohlava @michalkurka  I've run JMH for both Jama-based and netlib-java-based versions and opened a [pull request](https://github.com/h2oai/h2o-3/pull/1025). As for this PR, we can either accept it or I can add more JMH benchmarks based on JUnit tests for PCA. What do you think?

**Comment by mmalohlava** [comment]
@h2o-ops please test!

File: h2o-algos/src/jmh/java/hex/pca/PCAImputeMissingScoringBench.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    Options opt = new OptionsBuilder()
+        .include(PCAImputeMissingScoringBench.class.getSimpleName())
+        .build();
+
+    new Runner(opt).run();
+  }
+
+  @Setup(Level.Invocation)
+  public void setup() {
+    water.util.Log.setLogLevel("ERRR");
```
Comment (mathemage) [inline h2o-algos/src/jmh/java/hex/pca/PCAImputeMissingScoringBench.java]: @mmalohlava 
> we are outputting lot of stuff to console and file logs, which could "harm" benchmark measurements - feel free to setup lot4j log level to ERROR

What is better to be used here: `ERRR` or `FATAL`?

**Comment by abal5** [comment]
@h2o-ops please test!

**Comment by mathemage** [comment]
I have re-run JMH benchmarks again, after `git rebase master` from upstream. The results are not so stable:
```
Benchmark                                                          (dataSetCase)  Mode  Cnt    Score     Error  Units
PCAImputeMissingScoringBench.measureImputeMissingScoring                     N/A  avgt   10    0.338 ±   0.302  ms/op
PCAImputeMissingTrainingBench.measureImputeMissingTraining                   N/A  avgt   10    1.502 ±   0.289  ms/op
PCAWideDataSetsScoringBench.measureWideDataSetsBenchScoringCase                1  avgt   10    0.447 ±   0.717  ms/op
PCAWideDataSetsScoringBench.measureWideDataSetsBenchScoringCase                2  avgt   10    0.685 ±   1.563  ms/op
PCAWideDataSetsScoringBench.measureWideDataSetsBenchScoringCase                3  avgt   10    0.462 ±   0.778  ms/op
PCAWideDataSetsScoringBench.measureWideDataSetsBenchScoringCase                4  avgt   10    0.725 ±   1.443  ms/op
PCAWideDataSetsScoringBench.measureWideDataSetsBenchScoringCase                5  avgt   10    0.988 ±   0.509  ms/op
PCAWideDataSetsScoringBench.measureWideDataSetsBenchScoringCase                6  avgt   10    1.194 ±   1.312  ms/op
PCAWideDataSetsTrainingBench.measureWideDataSetsBenchTrainingCase              1  avgt   10    7.412 ±  13.393  ms/op
PCAWideDataSetsTrainingBench.measureWideDataSetsBenchTrainingCase              2  avgt   10    7.377 ±  14.891  ms/op
PCAWideDataSetsTrainingBench.measureWideDataSetsBenchTrainingCase              3  avgt   10    6.799 ±  11.629  ms/op
PCAWideDataSetsTrainingBench.measureWideDataSetsBenchTrainingCase              4  avgt   10    7.299 ±  11.560  ms/op
PCAWideDataSetsTrainingBench.measureWideDataSetsBenchTrainingCase              5  avgt   10  719.020 ±  47.550  ms/op
PCAWideDataSetsTrainingBench.measureWideDataSetsBenchTrainingCase              6  avgt   10  867.691 ± 769.523  ms/op
```
Especially, the error in the last test. Should we look into this, or can we merge?

@mmalohlava @abal5 Can you re-run tests again?

**Comment by abal5** [comment]
@h2o-ops please test!

**Comment by mmalohlava** [comment]
@h2o-ops please test!

**Comment by wendycwong** [comment]
Karel:

I have found the following three datasets that you can use for your benchmarking:

1. slowest (3 hours), bigdata/laptop/jira/la1s.wc.arff.txt.zip
2. Slow: bigdata/laptop/jira/re0.wc.arff.txt.zip
3. 10 minutes: bigdata/laptop/jira/rotterdam.csv.zip

Have fun with them!


Wendy

**Comment by wendycwong** [comment]
Running with dataset bigdata/laptop/jira/la1s.wc.arff.txt.zip off a laptop.  Time is in ms.

 Run with KarelTooBox:
10539190
10559385
10637884
10661748
10401644
10644001
10628831
10840295
10550188
10507291


Original Master
10385293
11021579
10790719
10806109
10801693
10789376
10858211
10849897
10884138
10770748

**Comment by mmalohlava** [comment]
@wendycwong each number is a single measurement? Was it inside the same JVM? or did you restarted JVM? Did you warm up the JVM?

**Comment by mathemage** [comment]
> @wendycwong each number is a single measurement? Was it inside the same JVM? or did you restarted JVM? Did you warm up the JVM?

That's what we discussed with @wendycwong over Slack. JMH seems more convenient to me, even just for measuring times to build models. I can benchmarks that would measure *just the build times* - well, actually I think these are exactly what my benchmarks do, they measure training time. 

But perhaps I can separate data pre-processing and post-processing parts, just time measure pure PCA training time. What do you think, @mmalohlava and @wendycwong ?

**Comment by mathemage** [comment]
Ok, I am merging this PR, it seems ripe enough to me :smile: I will add the JMH benchmarks with parameter `svdImplementation` in my other PR #1025 - this is because there are some non-trivial dependencies on implementations of various classes that only exist since #1025 

Thanks for all your remarks, @mmalohlava @tomasnykodym @wendycwong @abal5 @jakubhava and @arnocandel !

**Pull request merged**